### PR TITLE
NMS-12080: Add poller patterns

### PR DESCRIPTION
**Pull request by fooker**
This adds support for patterns in the poller configuration.

* JIRA: http://issues.opennms.org/browse/NMS-12080

### CODE REVIEW

**Comment by j-white** [comment]
Pretty cool. Look forward to trying this out.

**Comment by fooker** [comment]
I've fixed reloading of pollers with patterns an can confirm, that this also fixes https://issues.opennms.org/browse/NMS-9181.

**Comment by j-white** [comment]
I ran into the following exception when trying this:
```
2019-04-26 16:52:26,702 ERROR [Main] o.o.n.p.Poller: start: Failed to schedule existing interfaces
java.util.regex.PatternSyntaxException: Illegal repetition near index 4
HTTP-{port:[0-9]+}
    ^
        at java.util.regex.Pattern.error(Pattern.java:1957) ~[?:1.8.0_191]
        at java.util.regex.Pattern.closure(Pattern.java:3159) ~[?:1.8.0_191]
        at java.util.regex.Pattern.sequence(Pattern.java:2136) ~[?:1.8.0_191]
        at java.util.regex.Pattern.expr(Pattern.java:1998) ~[?:1.8.0_191]
        at java.util.regex.Pattern.compile(Pattern.java:1698) ~[?:1.8.0_191]
        at java.util.regex.Pattern.<init>(Pattern.java:1351) ~[?:1.8.0_191]
        at java.util.regex.Pattern.compile(Pattern.java:1028) ~[?:1.8.0_191]
        at java.util.regex.Pattern.matches(Pattern.java:1133) ~[?:1.8.0_191]
        at org.opennms.netmgt.config.PollerConfigManager.isServiceInPackageAndEnabled(PollerConfigManager.java:625) ~[opennms-config-25.0.0-SNAPSHOT.jar:?]
        at org.opennms.netmgt.poller.Poller.pollableServiceInPackage(Poller.java:612) ~[opennms-services-25.0.0-SNAPSHOT.jar:?]
        at org.opennms.netmgt.poller.Poller.findPackageForService(Poller.java:591) ~[opennms-services-25.0.0-SNAPSHOT.jar:?]
        at org.opennms.netmgt.poller.Poller.scheduleService(Poller.java:501) ~[opennms-services-25.0.0-SNAPSHOT.jar:?]
        at org.opennms.netmgt.poller.Poller.access$100(Poller.java:80) ~[opennms-services-25.0.0-SNAPSHOT.jar:?]
```

Service definition looks like:
```
      <service name="HTTP" interval="300000" user-defined="false" status="on" pattern="HTTP-{port:[0-9]+}" >
         <parameter key="retry" value="1"/>
         <parameter key="timeout" value="3000"/>
         <parameter key="port" value="${pattern:port|80}"/>
         <parameter key="url" value="/"/>
         <parameter key="rrd-repository" value="/home/jesse/git/opennms/target/opennms-25.0.0-SNAPSHOT/share/rrd/response"/>
         <parameter key="rrd-base-name" value="http-${pattern:port|80}"/>
         <parameter key="ds-name" value="http"/>
      </service>
```

I wonder if it we worthwhile to make the pattern field a pure regex, and use named capture groups like we do https://github.com/OpenNMS/opennms/pull/2469 to extract the named parameters. This would allow us to have optional matches too.

File: core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/MapScope.java [hunk 1]
```diff
@@ -49,4 +50,13 @@ public MapScope(final Map<ContextKey, String> values) {
     public Set<ContextKey> keys() {
         return this.values.keySet();
     }
+
+    public static MapScope singeContext(final String context, final Map<String, String> values) {
```
[the remark this answers is not in the document]
Comment (fooker) [reply core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/MapScope.java]: Fixed

**Comment by fooker** [comment]
@j-white I've stolen your code and switched to a pure RegEx implementation.

**Comment by fooker** [comment]
> @j-white I've stolen your code and switched to a pure RegEx implementation.

This makes it somewhat ugly, as the regex must be escaped to be a valid XML attribute...

**Comment by j-white** [comment]
Agreed. The regex in the XML attributes does look pretty ugly.

Maybe we can just wrap these in a CDATA tag?

File: core/lib/src/main/java/org/opennms/core/utils/RegexUtils.java [hunk 2]
```diff
@@ -50,8 +51,8 @@
      * @param pattern the pattern from which to extract the named capture groups
      * @return an ordered list of named capture group candidates
      */
-    public static List<String> getNamedCaptureGroupsFromPattern(String pattern) {
-        final List<String> namedGroups = new LinkedList<>();
+    public static Set<String> getNamedCaptureGroupsFromPattern(String pattern) {
+        final Set<String> namedGroups = Sets.newHashSet();
```
Comment (j-white) [inline core/lib/src/main/java/org/opennms/core/utils/RegexUtils.java]: Can we use a `LinkedHashSet` to preserve ordering?

File: core/lib/src/test/java/org/opennms/core/utils/RegexUtilsTest.java [hunk 3]
```diff
@@ -41,6 +42,6 @@ public void canParseNamedCaptureGroupsFromPattern() {
         assertThat(RegexUtils.getNamedCaptureGroupsFromPattern(""), hasSize(0));
         assertThat(RegexUtils.getNamedCaptureGroupsFromPattern("(?<user>.*)"), contains("user"));
         assertThat(RegexUtils.getNamedCaptureGroupsFromPattern("Node /(?<poolName>.*?)/(?<poolMember>\\S+) address (?<poolAddr>\\S+) monitor status down. .*\\(slot(?<slotNum>[0-9]+)\\)"),
-                contains("poolName", "poolMember", "poolAddr", "slotNum"));
+                containsInAnyOrder("poolName", "poolMember", "poolAddr", "slotNum"));
```
Comment (j-white) [inline core/lib/src/test/java/org/opennms/core/utils/RegexUtilsTest.java]: Can we revert this and keep the order deterministic?

File: features/poller/shell/src/main/java/org/opennms/netmgt/poller/shell/Test.java [hunk 4]
```diff
@@ -140,7 +142,15 @@ public MonitoredService doInTransaction(TransactionStatus status) {
...
         ServiceMonitor monitor = null;
         if (monitorClass == null) {
-            monitor = pollerConfig.getServiceMonitor(serviceName);
+            final Optional<Package.ServiceMatch> service = pkg.findService(serviceName);
+            if (!service.isPresent()) {
+                System.err.printf("Error: Service %s not defined%n", serviceName);
+                return null;
+            }
+
+            // TODO fooker: Print matching pattern fields...
```
Comment (j-white) [inline features/poller/shell/src/main/java/org/opennms/netmgt/poller/shell/Test.java]: TODO

File: features/poller/shell/src/main/java/org/opennms/netmgt/poller/shell/Test.java [hunk 5]
```diff
@@ -140,7 +142,15 @@ public MonitoredService doInTransaction(TransactionStatus status) {
 
         ServiceMonitor monitor = null;
         if (monitorClass == null) {
-            monitor = pollerConfig.getServiceMonitor(serviceName);
+            final Optional<Package.ServiceMatch> service = pkg.findService(serviceName);
+            if (!service.isPresent()) {
+                System.err.printf("Error: Service %s not defined%n", serviceName);
```
Comment (j-white) [inline features/poller/shell/src/main/java/org/opennms/netmgt/poller/shell/Test.java]: `%n`  intentional?

File: opennms-doc/guide-admin/src/asciidoc/text/service-assurance/polling-packages.adoc [hunk 6]
```diff
@@ -151,6 +151,33 @@ The _IP Interface_ and _Service_ page show which _Polling Package_ and _Service_
...
+[[ga-pollerd-packages-patterns]]
+==== Service Patterns
+
+Usually, the _Poller_ used to monitor a _Service_ is found by the matching the pollers name with the service name.
+In addition, a matching poller can be found if an additional element `pattern` is specified for the poller.
+If so, the poller is used for all services matching the _RegEx_ pattern, too.
+
+The _RegEx_ pattern allows to specify named capture groups.
+There can be multiple capture groups inside of a pattern, but each must have a unique name.
+Please not, that the _RegEx_ must be escaped or be wrapped in a CDATA-Tag inside the configuration XML to make it a valid property.
```
Comment (j-white) [inline opennms-doc/guide-admin/src/asciidoc/text/service-assurance/polling-packages.adoc]: Please note

File: features/poller/shell/src/main/java/org/opennms/netmgt/poller/shell/Test.java (hunk 5 above)
Comment (fooker) [reply features/poller/shell/src/main/java/org/opennms/netmgt/poller/shell/Test.java]: Yes, It's used every where else in the file, so I just keeped the style conistent.

**Pull request merged**